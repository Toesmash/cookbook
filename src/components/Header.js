import React from 'react';
import { Link } from 'react-router-dom';

const Header = () => (
  <header className='header'>
    <div className='main-container'>
      <div className='header__content'>
        <Link className='header__title' to="/"><h1>CookBook</h1></Link>
        <Link className='header__recipe' to="/create">Add Recipe</Link>
      </div>
    </div>
  </header>
);

export default Header;