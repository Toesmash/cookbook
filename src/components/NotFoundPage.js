import React from 'react';
import { Link } from 'react-router-dom';

const NotFound = () => (
  <div className='notfound__msg'>
    <h1>Page not found :(</h1>
    <p><Link to="/">Go home</Link></p>
  </div>
);

export default NotFound;