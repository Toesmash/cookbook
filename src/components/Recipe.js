import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { FaUtensils, FaStopwatch, FaThumbsDown, FaThumbsUp } from 'react-icons/fa';
import { removeRecipe, likeRecipe, dislikeRecipe } from '../redux/actions/recipes';



export class Recipe extends React.Component {
  onRemove = () => {
    this.props.removeRecipe(this.props.recipe.id);
    this.props.history.push('/');
  };
  onLike = () => {
    this.props.likeRecipe(this.props.recipe.id)
  };
  onDislike = () => {
    this.props.dislikeRecipe(this.props.recipe.id)
  };

  render() {
    const { title, ingredients, instructions, id, feeds, preparation, rating } = this.props.recipe;
    return (
      <div className='main-container'>
        <h2 className='recipe__title'>{title.toUpperCase()}</h2>
        <div className='recipe__info'>
          <span><FaUtensils />: {feeds} {feeds === 1 ? 'portion' : 'portions'}</span>
          <span><FaStopwatch />: {preparation} minutes</span>
        </div>
        <div className='recipe__content'>
          <div className='recipe__ingredients'>
            <h3>Ingredients</h3>
            {
              ingredients.map((item, index) => {
                return <p key={index}>- {item}</p>
              })
            }
          </div>
          <div className='recipe__instructions'>
            <h3>Instructions</h3>
            {
              instructions.map((item, index) => {
                return (
                  <div className='recipe__steps' key={index}>
                    <span className='recipe__index'>{index + 1}.</span>
                    <div>
                      {item}
                    </div>
                  </div>
                )
                // return <p key={index} className='recipe__step'><span className='recipe__index'>{index + 1}.</span> {item}</p>
              })
            }
          </div>

        </div>
        <div className='recipe__actions'>
          <div className='buttons'>
            <button className={rating.like ? 'button button--rating button--up button--like' : 'button button--rating button--up'} onClick={this.onLike}><FaThumbsUp /></button>
            <button className={rating.dislike ? 'button button--rating button--down button--dislike' : 'button button--rating button--down'} onClick={this.onDislike}><FaThumbsDown /></button>
          </div>
          <div className='buttons'>
            <Link to={`/edit/${id}`}><button className='button button--edit'>Edit Recipe</button></Link>
            <button className='button button--delete' onClick={this.onRemove} >Delete Recipe</button>
          </div>
        </div>
      </div>
    );
  };
};

const mapDispatchToProps = (dispatch) => ({
  removeRecipe: (id) => dispatch(removeRecipe(id)),
  likeRecipe: (id) => dispatch(likeRecipe(id)),
  dislikeRecipe: (id) => dispatch(dislikeRecipe(id))
});

const mapStateToProps = (store, props) => ({
  recipe: store.recipes.find((item) => {
    return item.id == props.match.params.id;
  })
});

export default connect(mapStateToProps, mapDispatchToProps)(Recipe);