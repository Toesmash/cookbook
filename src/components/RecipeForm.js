import React from 'react';
import * as Yup from 'yup';
import { withFormik, Form, Field } from 'formik';

export const RecipeForm = ({ errors, touched, values }) => {
  return (
    <div className='main-container'>
      <h2 className='form__title'>Recipe form</h2>
      <Form>
        <div className='form__basicinfo'>
          <div className='form__group'>
            <label>Title <span className='form__required'>*</span></label>
            <Field
              type="text"
              name="title"
              component="input"
              placeholder="Dumplings"
              className={errors.title && touched.title ? 'form__input form__input--error' : 'form__input'}
            />
            {errors.title && touched.title && <div className='input__feedback'>{errors.title}</div>}
          </div>

          <div className='form__group'>
            <label>Portions <span className='form__required'>*</span></label>
            <Field
              type="number"
              name="feeds"
              component="input"
              placeholder="4"
              className={errors.feeds && touched.feeds ? 'form__input form__input--error' : 'form__input'}
            />
            {errors.feeds && touched.feeds && <div className='input__feedback'>{errors.feeds}</div>}
          </div>

          <div className='form__group'>
            <label>Time needed (minutes) <span className='form__required'>*</span></label>
            <Field
              type="number"
              name="preparation"
              component="input"
              placeholder="60"
              className={errors.preparation && touched.preparation ? 'form__input form__input--error' : 'form__input'}
            />
            {errors.preparation && touched.preparation && <div className='input__feedback'>{errors.preparation}</div>}
          </div>
        </div>
        <div className='form__advancedinfo'>
          <div className='form__group form__group--ingredients'>
            <label>Ingredients <span className='form__required'>*</span></label>
            <Field
              type="text"
              name="ingredients"
              component="textarea"
              placeholder={`350g flour\n250ml water\n1tbs salt\n...`}
              className={errors.ingredients && touched.ingredients ? 'form__input form__input--textarea form__input--error ' : 'form__input form__input--textarea'}
            />
            {errors.ingredients && touched.ingredients && <div className='input__feedback'>{errors.ingredients}</div>}
          </div>

          <div className='form__group  form__group--instructions'>
            <label>Instructions <span className='form__required'>*</span></label>
            <Field
              type="text"
              name="instructions"
              component="textarea"
              placeholder={`Mix flour with water\nAdd salt\n...`}
              className={errors.instructions && touched.instructions ? 'form__input form__input--textarea form__input--error ' : 'form__input form__input--textarea'}
            />
            {errors.instructions && touched.instructions && <div className='input__feedback'>{errors.instructions}</div>}
          </div>
        </div>
        <div className='form__button'>
          <button className='button button--add' type='submit'>Submit recipe</button>
        </div>
      </Form>
    </div>
  );
}

const schema = {
  title: Yup.string().required('Required field').matches(/^[a-zA-Z0-9,\.\-\/:\(\)\s]*$/, 'No special characters allowed').max(16, 'Maximum allowed characters is 16').trim().lowercase(),
  feeds: Yup.number().required('Required field').min(1, 'Minimum number of portions is 1').max(8, 'Maximum number of portions is 8'),
  preparation: Yup.number().required('Required field').min(10, 'Minimum preparation time is 10minutes').max(300, 'Maximum preparation time is 5 hours'),
  ingredients: Yup.string().required('Required field').matches(/^[a-zA-Z0-9,\.\-\/:\(\)\n\s]*$/, 'No special characters allowed').trim(),
  instructions: Yup.string().required('Required field').matches(/^[a-zA-Z0-9,\.\-\/:\(\)\n\s]*$/, 'No special characters allowed').trim()
};

export const FormikRecipeForm = withFormik({
  mapPropsToValues({ recipe }) {
    return {
      title: recipe ? recipe.title : '',
      feeds: recipe ? recipe.feeds : '',
      preparation: recipe ? recipe.preparation : '',
      ingredients: recipe ? recipe.ingredients.join('\n') : '',
      instructions: recipe ? recipe.instructions.join('\n') : ''
    }
  },
  validationSchema: Yup.object().shape(schema),
  handleSubmit(values, { props }) {
    props.onSubmit({
      ...values,
      ingredients: values.ingredients.split('\n').filter(item => item),
      instructions: values.instructions.split('\n').filter(item => item)
    });
  }

})(RecipeForm);

export default FormikRecipeForm;

// export default class RecipeForm extends React.Component {
//   constructor(props) {
//     super(props);
//     this.state = {
//       title: props.recipe ? props.recipe.title : '',
//       feeds: props.recipe ? props.recipe.feeds : '',
//       preparation: props.recipe ? props.recipe.preparation : '',
//       ingredients: props.recipe ? props.recipe.ingredients : [],
//       instructions: props.recipe ? props.recipe.instructions : [],
//       error: ''
//     };
//   }

//   onTitleChange = (e) => {
//     const title = e.target.value;
//     this.setState(() => ({ title }))
//   }
//   onFeedsChange = (e) => {
//     const feeds = e.target.value;
//     this.setState(() => ({ feeds }))

//   }
//   onPrepChange = (e) => {
//     const preparation = e.target.value;
//     this.setState(() => ({ preparation }))

//   }
//   onInstructionChange = (e) => {
//     const instructions = e.target.value;
//     this.setState(() => ({ instructions: instructions.split("\n") }))
//   }
//   onIngredientsChange = (e) => {
//     const ingredients = e.target.value;
//     this.setState(() => ({ ingredients: ingredients.split("\n") }))
//   }

//   onSubmit = (e) => {
//     e.preventDefault();
//     if (!this.state.title || !this.state.ingredients || !this.state.instructions) {
//       this.setState(() => ({ error: 'Please set correct' }))
//     }
//     else {
//       this.setState(() => ({
//         error: ''
//       }))
//       this.props.onSubmit({
//         title: this.state.title,
//         feeds: this.state.feeds,
//         preparation: this.state.preparation,
//         ingredients: this.state.ingredients.filter((item) => (item)),
//         instructions: this.state.instructions.filter((item) => (item))
//       })
//     }
//   }

//   render() {
//     return (
//       <div>
//         {this.state.error && <p>{this.state.error}</p>}
//         <form onSubmit={this.onSubmit}>
//           <input
//             autoFocus
//             type="text"
//             name="title"
//             placeholder="Title"
//             value={this.state.title}
//             onChange={this.onTitleChange}
//           />
//           <input
//             type="text"
//             name="feeds"
//             placeholder="2"
//             value={this.state.feeds}
//             onChange={this.onFeedsChange}
//           />
//           <input
//             type="text"
//             name="preparation"
//             placeholder="60"
//             value={this.state.preparation}
//             onChange={this.onPrepChange}
//           />
//           <textarea
//             placeholder="Ingredients"
//             value={this.state.ingredients.join("\n")}
//             onChange={this.onIngredientsChange}
//           />
//           <textarea
//             placeholder="Instructions"
//             value={this.state.instructions.join("\n")}
//             onChange={this.onInstructionChange}
//           />
//           <button>Submit</button>
//         </form>
//       </div>
//     );
//   }
// }


