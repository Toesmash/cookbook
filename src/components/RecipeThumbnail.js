import React from 'react';
import { Link } from 'react-router-dom';
import { FaThumbsDown, FaThumbsUp } from 'react-icons/fa'

const RecipeThumbnail = ({ id, title, rating }) => {
    return (
        <Link className='list__item' to={`/view/${id}`}>
            {title.toUpperCase()}
            {rating.like && <span className='list__rating list__rating--up'><FaThumbsUp /></span>}
            {rating.dislike && <span className='list__rating list__rating--down'><FaThumbsDown /></span>}
        </Link>
    );
}

export default RecipeThumbnail;


