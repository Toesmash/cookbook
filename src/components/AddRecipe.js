import React from 'react';
import { connect } from 'react-redux';
import RecipeForm from './RecipeForm';
import { addRecipe } from '../redux/actions/recipes';

export class AddRecipe extends React.Component {
  onSubmit = (recipe) => {
    this.props.addRecipe(recipe);
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <RecipeForm
          onSubmit={this.onSubmit}
        />
      </div>
    );
  };
}

const mapDispatchToProps = (dispatch) => ({
  addRecipe: (recipe) => dispatch(addRecipe(recipe))
});

export default connect(undefined, mapDispatchToProps)(AddRecipe);