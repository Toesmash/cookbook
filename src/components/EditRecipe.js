import React from 'react';
import { connect } from 'react-redux';
import RecipeForm from './RecipeForm';
import { editRecipe } from '../redux/actions/recipes';

const Edit = (props) => {
  return (
    <div>
      <RecipeForm
        recipe={props.recipe}
        onSubmit={(recipe) => {
          props.dispatch(editRecipe(props.recipe.id, recipe));
          props.history.push('/')
        }}
      />
    </div>
  );
}

const mapStateToProps = (store, props) => ({
  recipe: store.recipes.find((item) => {
    return item.id == props.match.params.id;
  })
});

export default connect(mapStateToProps)(Edit);