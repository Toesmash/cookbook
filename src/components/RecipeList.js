import React from 'react';
import { connect } from 'react-redux';
import RecipeThumbnail from './RecipeThumbnail';


export const RecipeList = (props) => {
  return (
    <div className='main-container main-container--list'>
      <h2 className='list__title'>List of recipes</h2>
      <div className='list__grid'>
        {
          props.recipes.length === 0 ? <p className='list__empty'>No recipes</p> : (
            props.recipes.map((recipe) => (
              <RecipeThumbnail key={recipe.id} {...recipe} />))
          )
        }
      </div>
    </div>
  );
}

const mapStateToProps = (store) => ({
  recipes: store.recipes
});

export default connect(mapStateToProps)(RecipeList);