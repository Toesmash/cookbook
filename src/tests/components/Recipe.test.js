import React from 'react';
import { shallow } from 'enzyme';
import { Recipe } from '../../components/Recipe';
import dummyData from '../../redux/fixtures/dummyData';

let wrapper, history, removeRecipe;

beforeEach(() => {
  history = { push: jest.fn() };
  removeRecipe = jest.fn();
  wrapper = shallow(<Recipe removeRecipe={removeRecipe} history={history} recipe={dummyData[0]} />);
});

test('Render Recipe component', () => {
  expect(wrapper).toMatchSnapshot();
});

test('Test Recipe onRemove call', () => {
  wrapper.find('button').at(1).simulate('click');
  expect(removeRecipe).toHaveBeenLastCalledWith(dummyData[0].id);
  expect(history.push).toHaveBeenLastCalledWith('/');
});