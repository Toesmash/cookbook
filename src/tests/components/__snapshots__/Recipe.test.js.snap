// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Recipe component 1`] = `
<div
  className="main-container"
>
  <h2
    className="recipe__title"
  >
    POTATO DUMPLINGS
  </h2>
  <div
    className="recipe__info"
  >
    <span>
      <Component />
      : 
      2
       
      portions
    </span>
    <span>
      <Component />
      : 
      60
       minutes
    </span>
  </div>
  <div
    className="recipe__content"
  >
    <div
      className="recipe__ingredients"
    >
      <h3>
        Ingredients
      </h3>
      <p
        key="0"
      >
        - 
        2 medium potatoes
      </p>
      <p
        key="1"
      >
        - 
        1cup flour
      </p>
      <p
        key="2"
      >
        - 
        150gr bacon
      </p>
      <p
        key="3"
      >
        - 
        1tsp salt
      </p>
      <p
        key="4"
      >
        - 
        1/2 tbsp oil
      </p>
      <p
        key="5"
      >
        - 
        125g bryndza cheese
      </p>
      <p
        key="6"
      >
        - 
        50ml cream
      </p>
    </div>
    <div
      className="recipe__instructions"
    >
      <h3>
        Instructions
      </h3>
      <div
        className="recipe__steps"
        key="0"
      >
        <span
          className="recipe__index"
        >
          1
          .
        </span>
        <div>
          Grate the potatoes and place in a bowl
        </div>
      </div>
      <div
        className="recipe__steps"
        key="1"
      >
        <span
          className="recipe__index"
        >
          2
          .
        </span>
        <div>
          Mix flour and salt with potatoes until you get a thick, sticky dough
        </div>
      </div>
      <div
        className="recipe__steps"
        key="2"
      >
        <span
          className="recipe__index"
        >
          3
          .
        </span>
        <div>
          Chop bacon into small cubes and fry in 1/2 Tablespoon oil on low heat until crisp and brown. Meanwhile bring a large pot of water to a boil and add some salt to it
        </div>
      </div>
      <div
        className="recipe__steps"
        key="3"
      >
        <span
          className="recipe__index"
        >
          4
          .
        </span>
        <div>
          Test a small spoonful of the halusky mix in the boiling water. It should cook into a soft dumpling in 3-4 minutes. If the mix seem like its not holding together, add more flour to thicken
        </div>
      </div>
      <div
        className="recipe__steps"
        key="4"
      >
        <span
          className="recipe__index"
        >
          5
          .
        </span>
        <div>
          Drip several halusky dumplings in the boiling water at a time. They will sink to the bottom so give them a quick stir
        </div>
      </div>
      <div
        className="recipe__steps"
        key="5"
      >
        <span
          className="recipe__index"
        >
          6
          .
        </span>
        <div>
          Cook halusky until they float to the top and have changed color. Strain out with a slotted spoon
        </div>
      </div>
      <div
        className="recipe__steps"
        key="6"
      >
        <span
          className="recipe__index"
        >
          7
          .
        </span>
        <div>
          Transfer to a large bowl and mix with the cheese until evenly coated
        </div>
      </div>
      <div
        className="recipe__steps"
        key="7"
      >
        <span
          className="recipe__index"
        >
          8
          .
        </span>
        <div>
          Divide halusky onto plates and top with the fried bacon
        </div>
      </div>
      <div
        className="recipe__steps"
        key="8"
      >
        <span
          className="recipe__index"
        >
          9
          .
        </span>
        <div>
          Enjoy :)
        </div>
      </div>
    </div>
  </div>
  <div
    className="recipe__actions"
  >
    <div
      className="buttons"
    >
      <button
        className="button button--rating button--up button--like"
        onClick={[Function]}
      >
        <Component />
      </button>
      <button
        className="button button--rating button--down"
        onClick={[Function]}
      >
        <Component />
      </button>
    </div>
    <div
      className="buttons"
    >
      <Link
        replace={false}
        to="/edit/1cn0agbls"
      >
        <button
          className="button button--edit"
        >
          Edit Recipe
        </button>
      </Link>
      <button
        className="button button--delete"
        onClick={[Function]}
      >
        Delete Recipe
      </button>
    </div>
  </div>
</div>
`;
